marimo/_ast/compiler.py

```diff
@@ -91,17 +97,22 @@ def compile_cell(code: str, cell_id: CellId_t) -> CellImpl:
     else:
         expr = "None"
 
-    # store the cell's code in Python's linecache so debuggers can find it
-    body_filename = get_filename(cell_id)
-    last_expr_filename = get_filename(cell_id, suffix="_output")
-    # cache the entire cell's code
-    cache(body_filename, code)
-    if sys.version_info >= (3, 9):
-        # ast.unparse only available >= 3.9
-        cache(
-            last_expr_filename,
-            ast.unparse(expr) if not isinstance(expr, str) else "None",
-        )
+    if is_script:
+        ast.increment_lineno(module, start_line - 1)
+        body_filename = filename
+        last_expr_filename = filename
```

Comment:
The `ast.increment_lineno` gets the correct line when running the script and also when running with the debugger. The line column is off by 4 spaces. I haven't looked into updating it yet.

```
Traceback (most recent call last):
  File "/marimo/tests/_runtime/script_data/script_exception.py", line 15, in <module>
    app.run()
  File "/marimo/marimo/_ast/app.py", line 282, in run
    outputs, glbls = AppScriptRunner(InternalApp(self)).run()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/app/script_runner.py", line 82, in run
    outputs, defs = self._run_synchronous(
                    ^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/app/script_runner.py", line 111, in _run_synchronous
    output = execute_cell(cell, glbls, self.app.graph)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/executor.py", line 83, in execute_cell
    return EXECUTION_TYPES[execution_type].execute_cell(cell, glbls, graph)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/executor.py", line 137, in execute_cell
    exec(cell.body, glbls)
  File "/marimo/tests/_runtime/script_data/script_exception.py", line 10, in <module>
    y = y / x
    ^
NameError: name 'y' is not defined
```

I haven't looked into adding the source code. I'm not entirely sure what that looks like yet. I also wasn't sure if I correctly moved the caching code into the else-block.

Reply:
Hmm, not sure what you mean by the line column is off? This looks good to me

Reply:
I think the traceback is still referencing the column (4) from `__marimo__cell_Hbol_.py`
```
    y = y / x
    ^
```

I think it should be 4 chars to the right (column 8) for the original script.
```
    y = y / x
        ^
```

Reply:
Oh I see. Increment lineno is actually pretty naive

```python
def increment_lineno(node, n=1):
    """
    Increment the line number of each node in the tree starting at *node* by *n*.
    This is useful to "move code" to a different location in a file.
    """
    for child in walk(node):
        if 'lineno' in child._attributes:
            child.lineno = getattr(child, 'lineno', 0) + n
    return node
```

See https://github.com/python/cpython/blob/97248204a1cc10bd3a66ce857535d133f6bc86a1/Lib/ast.py#L260
for python's latest implementation.

I think:

```python
def fix_source_position(node, n=1, t=4):
    for child in ast.walk(node):
        # TypeIgnore is a special case where lineno is not an attribute
        # but rather a field of the node itself.
        # Note, TypeIgnore does not have a "col_offset"
        if isinstance(child, TypeIgnore):
            child.lineno = getattr(child, 'lineno', 0) + n
            continue

        if 'lineno' in child._attributes:
            child.lineno = getattr(child, 'lineno', 0) + n

        if 'col_offset' in child._attributes:
            child.col_offset = getattr(child, 'col_offset', 0) + t

        if (
            "end_lineno" in child._attributes
            and (end_lineno := getattr(child, "end_lineno", 0)) is not None
        ):
            child.end_lineno = end_lineno + n
            
        if (
            "end_col_offset" in child._attributes
            and (end_col_offset := getattr(child, "end_col_offset", 0)) is not None
        ):
            child.end_col_offset = end_col_offset + n
    return node
```

Reply:
Thanks for this code! I was able to add it in `compile.py`. However, I'm wondering if it should be more similar to [fix_missing_locations](https://github.com/python/cpython/blob/5505b91a684b0fc7ffcb3a5b325302671d74fb15/Lib/ast.py#L158), which uses the parent lineno if the child does not have a lineno attribute, rather than defaulting to 0?

Reply:
Or does it default to 0 because of these are cells?